### DISCUSSION TITLE
FEM challenge error visualization

### DISCUSSION THREAD

**Discussion by ignaciobartol** [seamplex/feenox#19] [Show and tell]
FEM challenge from Jeremy Theler @ [Linkedin](https://www.linkedin.com/feed/update/urn:li:activity:7216466748546400256/)
### Problem Statement

The problem was to approximate the integral of a quarter-circle considering the curved second-order `line3` element lying on the x-y plane defined by the corner nodes:

$$
\mathbf{x}_1 = [0, 1]
$$

$$
\mathbf{x}_2 = [1, 0]
$$

with the mid-edge node located at:

$$
\mathbf{x}_3 = \left[ \frac{1}{\sqrt{2}}, \frac{1}{\sqrt{2}} \right]
$$

Using the shape functions for $\xi \in [-1, 1]$:

$$
h_1(\xi) = \frac{\xi \cdot (\xi - 1)}{2}
$$

$$
h_2(\xi) = \frac{\xi \cdot (\xi + 1)}{2}
$$

$$
h_3(\xi) = (1 + \xi) \cdot (1 - \xi)
$$

compute the Lebesgue measure (i.e., length) of the element. The real length of the curve is $\pi/2 \approx 1.5708$. After solving the problem, we can plot the two curves to visualize the error, the three nodes are chosen to match the real curve.

## Mathematical Description / Solution

The position $\mathbf{x}(\xi)$ on the element can be described as:

$$
\mathbf{x}(\xi) = h_1(\xi) \cdot \mathbf{x}_1 + h_2(\xi) \cdot \mathbf{x}_2 + h_3(\xi) \cdot \mathbf{x}_3
$$

To find the length of the element, we need to integrate the arc length differential:

$$
L = \int_{-1}^{1} \left\| \frac{d\mathbf{x}(\xi)}{d\xi} \right\| d\xi
$$

First, we find the derivative of $\mathbf{x}(\xi)$ (this can be done in Feenox instead of finding them by hand, will update):

$$
\frac{d\mathbf{x}(\xi)}{d\xi} = \frac{d}{d\xi} \left( h_1(\xi) \cdot \mathbf{x}_1 + h_2(\xi) \cdot \mathbf{x}_2 + h_3(\xi) \cdot \mathbf{x}_3 \right)
$$

We can compute the derivatives of the shape functions:

$$
\frac{dh_1(\xi)}{d\xi} = \xi - \frac{1}{2}
$$

$$
\frac{dh_2(\xi)}{d\xi} = \xi + \frac{1}{2}
$$

$$
\frac{dh_3(\xi)}{d\xi} = -2\xi
$$

Therefore:

$$
\frac{d\mathbf{x}(\xi)}{d\xi} = \left( \xi - \frac{1}{2} \right) \cdot \mathbf{x}_1 + \left( \xi + \frac{1}{2} \right) \cdot \mathbf{x}_2 - 2\xi \cdot \mathbf{x}_3
$$

Substitute the node coordinates:

$$
\frac{d\mathbf{x}(\xi)}{d\xi} = \left( \xi - \frac{1}{2} \right) \cdot [0, 1] + \left( \xi + \frac{1}{2} \right) \cdot [1, 0] - 2\xi \cdot \left[ \frac{1}{\sqrt{2}}, \frac{1}{\sqrt{2}} \right]
$$

Simplifying:

$$
\frac{d\mathbf{x}(\xi)}{d\xi} = \left[ \left( \xi + \frac{1}{2} \right) - \frac{2\xi}{\sqrt{2}}, \left( \xi - \frac{1}{2} \right) - \frac{2\xi}{\sqrt{2}} \right]
$$

$$
\frac{d\mathbf{x}(\xi)}{d\xi} = \left[ \xi + \frac{1}{2} - \frac{2\xi}{\sqrt{2}}, \xi - \frac{1}{2} - \frac{2\xi}{\sqrt{2}} \right]
$$

Next, we compute the magnitude of this derivative:

$$
\left\| \frac{d\mathbf{x}(\xi)}{d\xi} \right\| = \sqrt{ \left( \xi + \frac{1}{2} - \frac{2\xi}{\sqrt{2}} \right)^2 + \left( \xi - \frac{1}{\sqrt{2}} - \frac{2\xi}{\sqrt{2}} \right)^2 }
$$

Finally, integrate this magnitude over $[-1, 1]$ to find the length $L$:

$$
L = \int_{-1}^{1} \sqrt{ \left( \xi + \frac{1}{2} - \frac{2\xi}{\sqrt{2}} \right)^2 + \left( \xi - \frac{1}{\sqrt{2}} - \frac{2\xi}{\sqrt{2}} \right)^2 } d\xi
$$

# FeenoX solution
```feenox
# Define the nodes and variables
VECTOR x1[2] DATA 0 1
VECTOR x2[2] DATA 1 0 
VECTOR x3[2] DATA 1/sqrt(2) 1/sqrt(2)
VAR xi
 
# Shape function derivatives
dh1_dxi(xi) = xi - 0.5
dh2_dxi(xi) = xi + 0.5
dh3_dxi(xi) = -2 * xi
 
# Derivative of x with respect to xi
dx_dxi_x(xi) = dh1_dxi(xi) * x1[1] + dh2_dxi(xi) * x2[1] + dh3_dxi(xi) * x3[1]
dx_dxi_y(xi) = dh1_dxi(xi) * x1[2] + dh2_dxi(xi) * x2[2] + dh3_dxi(xi) * x3[2]
 
integrand(xi) = sqrt(dx_dxi_x(xi)^2 + dx_dxi_y(xi)^2)
 
# Integration limits
a = -1
b = 1
 
# Compute the length by numerical integration
length = integral(integrand(xi), xi, a, b)
 
# Output the computed length
print length
```

The length will be ~1.56242 using this method.

```python
import numpy as np
import matplotlib.pyplot as plt

# Position function x(xi) based on the shape functions
def x_xi(xi):
    h1 = 0.5 * xi * (xi - 1)
    h2 = 0.5 * xi * (xi + 1)
    h3 = 1 - xi**2
    return h1 * x1 + h2 * x2 + h3 * x3

# Generate xi values and corresponding x(xi) values
xi_vals = np.linspace(-1, 1, 100)
x_vals = np.array([x_xi(xi) for xi in xi_vals])

# Generate points for the quarter circle
theta = np.linspace(0, np.pi/2, 100)
quarter_circle_x = np.cos(theta)
quarter_circle_y = np.sin(theta)

# Plot the curve defined by x(ξ) and the quarter circle
plt.plot(x_vals[:, 0], x_vals[:, 1], label=r'x($\xi$)', linestyle='--')
plt.plot(quarter_circle_x, quarter_circle_y, label='Quarter Circle', linestyle=':')
plt.scatter([x1[0], x2[0], x3[0]], [x1[1], x2[1], x3[1]], color='red', label='Nodes')
plt.xlabel('x')
plt.ylabel('y')
plt.title(r'Curve Defined by x($\xi$) and Quarter Circle')
plt.legend()
plt.grid(True)
plt.axis('equal')
plt.show()
```

![plot](https://github.com/user-attachments/assets/416047e1-321e-4862-b599-5d069f632fdc)

**Comment by gtheler** [comment]
Here is a cleaned-up version:

```
# based on https://github.com/seamplex/feenox/discussions/19
# originally by ibartol

# Define the nodes and variables
VECTOR x1[2] DATA 0 1
VECTOR x2[2] DATA 1 0 
VECTOR x3[2] DATA 1/sqrt(2) 1/sqrt(2)

h1(xi) = xi*(xi-1)/2
h2(xi) = xi*(xi+1)/2
h3(xi) = (1+xi)*(1-xi)

# Shape function derivatives
VAR xi'
h1'(xi) = derivative(h1(xi'), xi', xi)
h2'(xi) = derivative(h2(xi'), xi', xi)
h3'(xi) = derivative(h3(xi'), xi', xi)
 
# Derivative of x & y with respect to xi
dx_dxi(xi) = h1'(xi)*x1[1] + h2'(xi)*x2[1] + h3'(xi)*x3[1]
dy_dxi(xi) = h1'(xi)*x1[2] + h2'(xi)*x2[2] + h3'(xi)*x3[2] 


PRINT integral(sqrt(dx_dxi(xi)^2+dy_dxi(xi)^2),xi,-1,1)
```

Also, see https://github.com/seamplex/feenox/blob/main/tests/single-arc.fee

**Reply by ignaciobartol** [reply]
That's so neat. Thanks!
